File: factory_flasher/internal/layout/layout.go
// Package layout is the USB factory flash map.
//
// Must stay aligned with server/factory_layout.py, scripts/hw_restore.sh,
// and scripts/snapshot_factory.sh.
package layout

import (
	"fmt"
	"os"
	"path/filepath"
)

const (
	Chip      = "esp32s3"
	Baud      = 460800
	Before    = "default_reset"
	After     = "hard_reset"
	FlashMode = "dio"
	FlashFreq = "80m"
	FlashSize = "4MB"
)

// Espressif USB-JTAG/serial, Silicon Labs CP210x, WCH CH340.
var USBVendorIDs = []int{0x303A, 0x10C4, 0x1A86}

type File struct {
	Key      string
	Offset   uint32
	Filename string
	BuildRel string
}

var Files = []File{
	{Key: "bootloader", Offset: 0x0, Filename: "bootloader.bin", BuildRel: "bootloader/bootloader.bin"},
	{Key: "partitions", Offset: 0x8000, Filename: "partition-table.bin", BuildRel: "partition_table/partition-table.bin"},
	{Key: "otadata", Offset: 0xF000, Filename: "ota_data_initial.bin", BuildRel: "ota_data_initial.bin"},
	{Key: "app", Offset: 0x20000, Filename: "app.bin", BuildRel: "searaboom.bin"},
	{Key: "storage", Offset: 0x3A0000, Filename: "storage.bin", BuildRel: "storage.bin"},
}

func RequiredFilenames() []string {
	out := make([]string, len(Files))
	for i, f := range Files {
		out[i] = f.Filename
	}
	return out
}

func OffsetHex(offset uint32) string {
	return fmt.Sprintf("0x%x", offset)
}

func SlotReady(dir string) bool {
	for _, f := range Files {
		st, err := os.Stat(filepath.Join(dir, f.Filename))
		if err != nil || st.IsDir() {
			return false
		}
	}
	return true
}

func FileMap(dir string) (map[string]string, error) {
	out := make(map[string]string, len(Files))
	var missing []string
	for _, f := range Files {
		path := filepath.Join(dir, f.Filename)
		st, err := os.Stat(path)
		if err != nil || st.IsDir() {
			missing = append(missing, f.Filename)
			continue
		}
		out[f.Filename] = path
	}
	if len(missing) > 0 {
		return nil, fmt.Errorf("missing factory files: %s", joinComma(missing))
	}
	return out, nil
}

func joinComma(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += ", "
		}
		out += p
	}
	return out
}

File: factory_flasher/layout.py
"""USB factory flash layout.

Must stay aligned with firmware/build/flasher_args.json, scripts/hw_restore.sh,
and scripts/snapshot_factory.sh. The desktop flasher and the server both import this.
"""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

CHIP = "esp32s3"
BAUD = 460800
BEFORE = "default_reset"
AFTER = "hard_reset"
FLASH_MODE = "dio"
FLASH_FREQ = "80m"
FLASH_SIZE = "4MB"

# Espressif USB-JTAG/serial, Silicon Labs CP210x, WCH CH340 (see udev rules).
USB_VENDOR_IDS = (0x303A, 0x10C4, 0x1A86)


@dataclass(frozen=True)
class FactoryFile:
    key: str
    offset: int
    filename: str
    build_rel: str


FACTORY_FILES: tuple[FactoryFile, ...] = (
    FactoryFile("bootloader", 0x0, "bootloader.bin", "bootloader/bootloader.bin"),
    FactoryFile("partitions", 0x8000, "partition-table.bin", "partition_table/partition-table.bin"),
    FactoryFile("otadata", 0xF000, "ota_data_initial.bin", "ota_data_initial.bin"),
    FactoryFile("app", 0x20000, "app.bin", "searaboom.bin"),
    FactoryFile("storage", 0x3A0000, "storage.bin", "storage.bin"),
)

REQUIRED_FILENAMES = tuple(item.filename for item in FACTORY_FILES)


def factory_plan(build_dir: Path | None = None) -> list[dict]:
    plan: list[dict] = []
    for item in FACTORY_FILES:
        rec: dict = {
            "key": item.key,
            "offset": item.offset,
            "filename": item.filename,
        }
        if build_dir is not None:
            rec["build"] = Path(build_dir) / item.build_rel
        plan.append(rec)
    return plan


def offset_hex(offset: int) -> str:
    return hex(offset)


def esptool_write_args(port: str, file_paths: dict[str, Path]) -> list[str]:
    """Args after `python -m esptool`. Matches scripts/hw_restore.sh."""
    missing = [name for name in REQUIRED_FILENAMES if name not in file_paths]
    if missing:
        raise FileNotFoundError(", ".join(missing))
    cmd = [
        "--chip", CHIP,
        "-p", port,
        "-b", str(BAUD),
        "--before", BEFORE,
        "--after", AFTER,
        "write_flash",
        "--flash_mode", FLASH_MODE,
        "--flash_freq", FLASH_FREQ,
        "--flash_size", FLASH_SIZE,
    ]
    for item in FACTORY_FILES:
        path = Path(file_paths[item.filename])
        if not path.is_file():
            raise FileNotFoundError(item.filename)
        cmd.extend([offset_hex(item.offset), str(path)])
    return cmd


def slot_ready(directory: Path) -> bool:
    return all((Path(directory) / name).is_file() for name in REQUIRED_FILENAMES)


def file_map(directory: Path) -> dict[str, Path]:
    directory = Path(directory)
    mapping = {name: directory / name for name in REQUIRED_FILENAMES}
    missing = [name for name, path in mapping.items() if not path.is_file()]
    if missing:
        raise FileNotFoundError(", ".join(missing))
    return mapping
